src/libmongoc/src/mongoc/mongoc-topology-description-apm.c

```diff
@@ -148,38 +174,52 @@ _mongoc_topology_description_monitor_opening (mongoc_topology_description_t *td)
 /* TopologyDescriptionChangedEvent */
 void
 _mongoc_topology_description_monitor_changed (const mongoc_topology_description_t *prev_td,
-                                              const mongoc_topology_description_t *new_td)
+                                              const mongoc_topology_description_t *new_td,
+                                              const mongoc_log_and_monitor_instance_t *log_and_monitor)
 {
-   if (new_td->apm_callbacks.topology_changed) {
+   mongoc_structured_log (log_and_monitor->structured_log,
+                          MONGOC_STRUCTURED_LOG_LEVEL_DEBUG,
+                          MONGOC_STRUCTURED_LOG_COMPONENT_TOPOLOGY,
+                          "Topology description changed",
+                          oid ("topologyId", &new_td->topology_id),
+                          topology_description_as_json ("previousDescription", prev_td),
+                          topology_description_as_json ("newDescription", new_td));
+
+   if (log_and_monitor->apm_callbacks.topology_changed) {
       mongoc_apm_topology_changed_t event;
 
       /* callbacks, context, and id are the same in previous and new td */
       bson_oid_copy (&new_td->topology_id, &event.topology_id);
-      event.context = new_td->apm_context;
+      event.context = log_and_monitor->apm_context;
       event.previous_description = prev_td;
       event.new_description = new_td;
 
-      new_td->apm_callbacks.topology_changed (&event);
+      log_and_monitor->apm_callbacks.topology_changed (&event);
    }
 }
 
 /* TopologyClosedEvent */
 void
-_mongoc_topology_description_monitor_closed (const mongoc_topology_description_t *td)
+_mongoc_topology_description_monitor_closed (const mongoc_topology_description_t *td,
+                                             const mongoc_log_and_monitor_instance_t *log_and_monitor)
 {
-   if (td->apm_callbacks.topology_closed) {
-      mongoc_apm_topology_closed_t event;
+   BSON_ASSERT (td->type == MONGOC_TOPOLOGY_UNKNOWN);
+   BSON_ASSERT (mc_tpld_servers_const (td)->items_len == 0);
```

Comment:
Suggest adding a comment referencing the function/code responsible for ensuring these conditions; these assertions feel somewhat out-of-place otherwise (beyond the responsibility of merely logging the event on request).